infrastructure/src/main/java/org/corfudb/infrastructure/health/HealthStatus.java

```diff
@@ -0,0 +1,108 @@
+    @Getter
+    private final Set<Issue> runtimeHealthIssues;
+
+    private InitStatus initStatus = InitStatus.UNKNOWN;
+
+    public HealthStatus() {
+        this.initHealthIssues = new LinkedHashSet<>();
+        this.runtimeHealthIssues = new LinkedHashSet<>();
+    }
+
+    /**
+     * Add init health issue -- health monitor now expects the component to be initialized.
+     * If component has been not been previously initialized (UNKNOWN) or initialized (INITIALIZED),
+     * adding init issue will move it to INITIALIZED, all runtime issues are cleared.
+     * @param issue An init issue
+     */
+    public void addInitHealthIssue(Issue issue) {
+        // If component has been not been previously initialized (UNKNOWN) or
+        // initialized (INITIALIZED), adding init issue will move it to INITIALIZED,
+        // all runtime issues are cleared
+        if (initStatus == InitStatus.UNKNOWN || initStatus == InitStatus.INITIALIZED) {
+            initStatus = InitStatus.NOT_INITIALIZED;
+            runtimeHealthIssues.clear();
+        }
+        initHealthIssues.add(issue);
+    }
+
+    /**
+     * Remove the init health issue -- report to health monitor that the component is initialized.
+     * @param issue An init issue
+     */
+    public void resolveInitHealthIssue(Issue issue) {
+        if (initStatus == InitStatus.NOT_INITIALIZED) {
+            initStatus = InitStatus.INITIALIZED;
+        }
+        initHealthIssues.remove(issue);
+    }
+
+    /**
+     * Add runtime health issue -- health monitor now expects the component's runtime issue to recover.
+     * If component has been not been previously initialized (INITIALIZED), adding the runtime issue will
+     * result in exception.
+     * @param issue An init issue
+     */
+    public void addRuntimeHealthIssue(Issue issue) {
+        if (initStatus != InitStatus.INITIALIZED) {
+            throw new IllegalStateException("Runtime health issue can only be reported if the component is initialized");
+        }
+        runtimeHealthIssues.add(issue);
+    }
+
+    /**
+     * Remove the runtime health issue -- report to health monitor that this issue is resolved.
+     * @param issue An init issue
+     */
+    public void resolveRuntimeHealthIssue(Issue issue) {
+        runtimeHealthIssues.remove(issue);
+    }
+
+    public boolean isInitHealthy() {
+        return initHealthIssues.isEmpty() && initStatus == InitStatus.INITIALIZED;
+    }
+
+    public boolean isRuntimeHealthy() {
+        return runtimeHealthIssues.isEmpty() && initStatus == InitStatus.INITIALIZED;
+    }
+
+    /**
+     * Get the last reported runtime issue
+     * @return A possible runtime health issue
+     */
+    public Optional<Issue> getLatestRuntimeIssue() {
+        if (runtimeHealthIssues.isEmpty()) {
+            return Optional.empty();
+        }
+        return runtimeHealthIssues.stream().skip(runtimeHealthIssues.size() - 1).findFirst();
+    }
```

Comment:
Issues are unique, so we just want to report the latest unique issue. Here in the code example if we add the failedTask issue again, it's essentially a noop because it's the same issue that is still not resolved and is sitting in the LinkedHashSet, so no point in reporting it. If the unresponsive issue is resolved, the failedTask will be shown anyway next (because it's the first one that has appeared). I can rename this method to getLatestUniqueRuntimeIssue() for clarity.

Reply:
Got it - I don't think renaming it is necessary, maybe just a comment describing this behaviour. It can also make for a good test case.